onnxruntime/core/graph/graph.cc

```diff
@@ -4112,47 +4188,53 @@ ONNX_NAMESPACE::GraphProto Graph::ToGraphProtoWithExternalInitializers(const std
   return result;
 }
 
-void Graph::ToGraphProtoInternal(ONNX_NAMESPACE::GraphProto& graph_proto) const {
-  graph_proto_->clear_node();
-  graph_proto_->clear_input();
-  graph_proto_->clear_output();
-  graph_proto_->clear_value_info();
-  graph_proto.set_name(Name());
-  graph_proto.set_doc_string(Description());
+template <typename TInstance>
+void Graph::ToGraphProtoInternal(TInstance& instance, ONNX_NAMESPACE::GraphProto& graph_proto) {
```

Comment:
Would be good to avoid templatizing here as well.

One option would be to pass in `std::variant<Graph*, const Graph*>`. Use the mutable one if present and call ToProtoFinal.

Another would be to pass in a functor to do the node update. It could capture the const/non-const Graph instance and call Node::ToProto[Final] as applicable.